Analise Dados

In [69]:
#Importando as bibliotecas
import numpy as np
import pandas as pd

In [70]:
#Lendo os arquivos
pd.set_option('display.max_columns', None)

#leitura dos arquivos csv
df_agencias = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\agencias.csv")
df_clientes = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\clientes.csv")
df_colaborador_agencia = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\colaborador_agencia.csv")
df_colaborador = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\colaboradores.csv")
df_contas = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\contas.csv")
df_propostas_credito = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\propostas_credito.csv")
df_transacoes = pd.read_csv(r"C:\Users\UC396QV\OneDrive - EY\Desktop\Lighthouse\analise de dados\banvic_data\transacoes.csv")

Agencias por estado

In [71]:
estado = df_agencias.groupby('uf')['uf'].count()
print(estado)

uf
PE    1
RJ    1
RS    1
SC    1
SP    6
Name: uf, dtype: int64


Obtendo clientes por estado
Nesse caso sempre a sigla do estado eram os 2 últimos caracteres da coluna 'endereço'
Então foi feito tratamento para obter esses dados em uma nova coluna e foi feito seu agrupamento

In [72]:
df_clientes['uf'] = df_clientes['endereco'].str[-2:]
#print(df_clientes)

clientes_estado = df_clientes.groupby('uf')['uf'].count().sort_values(ascending=False)
print(clientes_estado)

uf
AM    53
ES    49
SP    48
MS    47
MA    44
CE    43
RR    41
AL    40
MG    40
AC    38
SC    37
PB    37
TO    37
RS    35
AP    35
GO    34
RN    34
BA    34
PR    33
DF    33
PA    33
RJ    32
PI    31
SE    29
MT    28
PE    28
RO    25
Name: uf, dtype: int64


Para agrupar por ano de nascimento foi feito lógica semelhante
Com a diferença de que o ano de nascimento eram os 4 primeiros caracterers do campo 'data_nascimento'

In [73]:
df_clientes['ano_nascimento'] = df_clientes['data_nascimento'].str[:4]
ano_nasc_clientes = df_clientes.groupby('ano_nascimento')['ano_nascimento'].count()
#clientes_nascimento.columns = ["index","ano"]
#clientes_nascimento.groupby('data_nascimento')['data_nascimento'].count().sort_values(ascending=False)
#ano = clientes_nascimento.groupby(clientes_nascimento)
print(ano_nasc_clientes)

ano_nascimento
1942    10
1943    16
1944    16
1945    13
1946    15
        ..
2003    16
2004     8
2005    19
2006    17
2007     6
Name: ano_nascimento, Length: 66, dtype: int64


In [74]:
A contagem por geração de idade foi feita manualmente, mas poderia também ser feita via código

SyntaxError: invalid syntax (28035348.py, line 1)

Obtenção de tipo de cliente (PJ ou PF), realizando agrupamento pela coluna 'tipo_cliente'

In [75]:
tipo_cliente = df_clientes.groupby('tipo_cliente')['tipo_cliente'].count().sort_values(ascending=False)
print(tipo_cliente)

tipo_cliente
PF    998
Name: tipo_cliente, dtype: int64


Agora focando nos colaboradores, foi feito agrupamento por agência

In [76]:
colaborador = df_colaborador_agencia.groupby('cod_agencia')['cod_agencia'].count()
print(colaborador)

cod_agencia
1     16
2     13
3     11
4      4
5     13
6     11
7      7
8      7
9      8
10    10
Name: cod_agencia, dtype: int64


Clientes por colaborador

In [77]:
clie_por_colab = df_contas.groupby('cod_colaborador')['num_conta'].count().sort_values(ascending=False)
print(clie_por_colab)

cod_colaborador
26     84
20     69
84     67
64     62
48     61
       ..
18      1
95      1
44      1
74      1
100     1
Name: num_conta, Length: 94, dtype: int64


Agrupamento por ano de nascimento do colaborador

In [78]:
df_colaborador['ano_nascimento'] = df_colaborador['data_nascimento'].str[:4]
ano_nasc_colaborador = df_colaborador.groupby('ano_nascimento')['ano_nascimento'].count()
print(ano_nasc_colaborador)

ano_nascimento
1962    1
1963    6
1964    3
1965    1
1966    3
1967    3
1968    2
1969    3
1970    4
1971    1
1973    4
1974    3
1975    4
1976    3
1977    5
1978    4
1979    1
1980    3
1981    3
1982    2
1983    1
1985    5
1986    3
1987    1
1988    1
1989    2
1990    3
1991    5
1992    2
1993    5
1994    4
1995    2
1996    3
1997    3
1998    1
Name: ano_nascimento, dtype: int64


Realizando merge dos datasets de colaborador e colaborador_agencia usando o codigo de colaborador como chave comum a ambas
Para obter o estador por colaborador

Adicionando coluna de 'uf' para abrigar a sigla do estado de cada colaborador

In [79]:
df_colaborador['uf'] = df_colaborador['endereco'].str[-2:]

In [80]:
df_colaboradores = df_colaborador_agencia.merge(df_colaborador[['cod_colaborador','ano_nascimento','uf']],on='cod_colaborador',how='left')
colaborador_uf = df_colaboradores.groupby('uf')['uf'].count().sort_values(ascending=False)
print(colaborador_uf)

uf
PE    8
AM    7
SC    7
SE    6
MS    6
RJ    5
RR    5
CE    5
RN    4
MT    4
MA    4
PB    3
TO    3
RO    3
GO    3
ES    3
DF    3
BA    3
AP    3
SP    3
AC    2
PR    2
PI    2
MG    2
PA    2
RS    1
AL    1
Name: uf, dtype: int64


# 3.1 Trimestre com mais transações

Dado que a data das transações está no formato YYYY-MM-DD, podemos presumir sempre que a data será 2 números delimitados por '-' e ','
Sendo o mês obtido da seguinte maneira

Adicionar uma coluna que ficará apenas com os meses

In [81]:
Porém antes é importante observar se a coluna de data está no formato correto
No caso estava no tipo objeto, sendo necessário realizar a conversão para tipo datetime
Como não havia formato padronizado para os microsegundos, o format foi colocado como 'mixed' onde a inferência é feita em execução, sendo mais demorado porém mais preciso

SyntaxError: invalid syntax (4165257599.py, line 1)

In [82]:
df_transacoes['data_transacao'] = pd.to_datetime(df_transacoes['data_transacao'], format='mixed')
print((df_transacoes))

       cod_transacao  num_conta                   data_transacao  \
0               3100        256        2020-05-22 03:15:21+00:00   
1              28405        256        2021-06-29 11:52:38+00:00   
2               1776        512        2020-05-17 13:04:50+00:00   
3              14311        512        2020-09-06 02:55:34+00:00   
4              25012        512        2020-10-16 09:26:52+00:00   
...              ...        ...                              ...   
71994          17007        764 2022-12-29 23:59:59.746608+00:00   
71995          22320        764 2022-12-29 23:59:59.055028+00:00   
71996           8697        509        2012-05-07 13:45:41+00:00   
71997          11574        766        2015-09-19 05:13:07+00:00   
71998          64396        766 2022-12-29 23:59:59.607373+00:00   

                         nome_transacao  valor_transacao  
0                                 Saque           -130.0  
1                                 Saque           -255.0  
2     

In [83]:
print((df_transacoes.dtypes))

cod_transacao                    int64
num_conta                        int64
data_transacao     datetime64[ns, UTC]
nome_transacao                  object
valor_transacao                float64
dtype: object


Com o valor devidamente obtido, agora é feito a extração do mês

In [84]:
df_transacoes['mes'] = df_transacoes['data_transacao'].dt.month
print(df_transacoes)

       cod_transacao  num_conta                   data_transacao  \
0               3100        256        2020-05-22 03:15:21+00:00   
1              28405        256        2021-06-29 11:52:38+00:00   
2               1776        512        2020-05-17 13:04:50+00:00   
3              14311        512        2020-09-06 02:55:34+00:00   
4              25012        512        2020-10-16 09:26:52+00:00   
...              ...        ...                              ...   
71994          17007        764 2022-12-29 23:59:59.746608+00:00   
71995          22320        764 2022-12-29 23:59:59.055028+00:00   
71996           8697        509        2012-05-07 13:45:41+00:00   
71997          11574        766        2015-09-19 05:13:07+00:00   
71998          64396        766 2022-12-29 23:59:59.607373+00:00   

                         nome_transacao  valor_transacao  mes  
0                                 Saque           -130.0    5  
1                                 Saque           -255.

Realizando o agrupamento por trimestre, levando em conta as movimentações feitas

Primeiro convertendo a coluna de valores negativos para positivos

In [85]:
df_transacoes['valor_transacao'] = df_transacoes['valor_transacao'].abs()

E por fim adicionando uma coluna de trimestre

In [86]:
df_transacoes['trimestre'] = df_transacoes['data_transacao'].dt.to_period('Q')

C:\Users\UC396QV\AppData\Local\Temp\ipykernel_8196\1191444974.py:1: UserWarning: Converting to PeriodArray/Index representation will drop timezone information.
  df_transacoes['trimestre'] = df_transacoes['data_transacao'].dt.to_period('Q')


In [87]:
print(df_transacoes)

       cod_transacao  num_conta                   data_transacao  \
0               3100        256        2020-05-22 03:15:21+00:00   
1              28405        256        2021-06-29 11:52:38+00:00   
2               1776        512        2020-05-17 13:04:50+00:00   
3              14311        512        2020-09-06 02:55:34+00:00   
4              25012        512        2020-10-16 09:26:52+00:00   
...              ...        ...                              ...   
71994          17007        764 2022-12-29 23:59:59.746608+00:00   
71995          22320        764 2022-12-29 23:59:59.055028+00:00   
71996           8697        509        2012-05-07 13:45:41+00:00   
71997          11574        766        2015-09-19 05:13:07+00:00   
71998          64396        766 2022-12-29 23:59:59.607373+00:00   

                         nome_transacao  valor_transacao  mes trimestre  
0                                 Saque            130.0    5    2020Q2  
1                                 S

Agora converter a coluna trimestre de objeto para inteiro

In [95]:
df_transacoes['trimestre'] = df_transacoes['trimestre'].astype(str)
df_transacoes['trimestre_num'] = df_transacoes['trimestre'].str[-1:]
print(df_transacoes)

       cod_transacao  num_conta                   data_transacao  \
0               3100        256        2020-05-22 03:15:21+00:00   
1              28405        256        2021-06-29 11:52:38+00:00   
2               1776        512        2020-05-17 13:04:50+00:00   
3              14311        512        2020-09-06 02:55:34+00:00   
4              25012        512        2020-10-16 09:26:52+00:00   
...              ...        ...                              ...   
71994          17007        764 2022-12-29 23:59:59.746608+00:00   
71995          22320        764 2022-12-29 23:59:59.055028+00:00   
71996           8697        509        2012-05-07 13:45:41+00:00   
71997          11574        766        2015-09-19 05:13:07+00:00   
71998          64396        766 2022-12-29 23:59:59.607373+00:00   

                         nome_transacao  valor_transacao  mes trimestre  \
0                                 Saque            130.0    5    2020Q2   
1                                

In [97]:
df_transacoes['trimestre_num'] = df_transacoes['trimestre_num'].astype(np.int64)

In [102]:
print(df_transacoes)

       cod_transacao  num_conta                   data_transacao  \
0               3100        256        2020-05-22 03:15:21+00:00   
1              28405        256        2021-06-29 11:52:38+00:00   
2               1776        512        2020-05-17 13:04:50+00:00   
3              14311        512        2020-09-06 02:55:34+00:00   
4              25012        512        2020-10-16 09:26:52+00:00   
...              ...        ...                              ...   
71994          17007        764 2022-12-29 23:59:59.746608+00:00   
71995          22320        764 2022-12-29 23:59:59.055028+00:00   
71996           8697        509        2012-05-07 13:45:41+00:00   
71997          11574        766        2015-09-19 05:13:07+00:00   
71998          64396        766 2022-12-29 23:59:59.607373+00:00   

                         nome_transacao  valor_transacao  mes trimestre  \
0                                 Saque            130.0    5    2020Q2   
1                                

Por fim, o agrupamento das movimentações por trimestre

In [106]:
movimentacoes_trimestre = df_transacoes.groupby('trimestre_num')['valor_transacao'].sum()
print(movimentacoes_trimestre)

trimestre_num
1     9660497.99
2     8637318.23
3    11720867.02
4    28121869.34
Name: valor_transacao, dtype: float64


O trimestre com o maior volume de movimentação foi o 4

Agora para obter a maior quantidade de transações

In [108]:
movimentacoes_trimestre = df_transacoes.groupby('trimestre_num').size()
print(movimentacoes_trimestre)

trimestre_num
1    10813
2    10991
3    12190
4    38005
dtype: int64


O trimestre com mais movimentações também foi o 4, com 38005 transações

Quanto a padronização de transações de meses com R ou sem, como já temos os meses em números, será utilizado da seguinte forma:
1 - Janeiro (Conta)
2 - Fevereiro (Conta)
3 - Março (Conta)
4 - Abril (Conta)
5 - Maio (Não conta)
6 - Junho (Não conta)
7 - Julho (Não conta)
8 - Agosto (Não conta)
9 - Setembro (Conta)
10 - Outubro (Conta)
11 - Novembro (Conta) 
12 - Dezembro (Conta)

Ou seja, apenas 4 meses estão divergentes

Analisando os padrões nos meses de 1 a 4, e 9 a 12, temos:

In [110]:
#Analisando os meses de janeiro a abril
#criando dataset para os meses
df_mov_janeiro_abril = df_transacoes[(df_transacoes['mes'] >=1) & (df_transacoes['mes'] <=4)]
df_mov_setembro_dezembro = df_transacoes[(df_transacoes['mes'] >=9) & (df_transacoes['mes'] <=12)]
df_mov_maio_agosto = df_transacoes[(df_transacoes['mes'] >=5) & (df_transacoes['mes'] <=8)]

#agrupando
res_mov_janeiro_abril = df_mov_janeiro_abril.groupby('nome_transacao')['valor_transacao'].sum()
res_mov_setembro_dezembro = df_mov_setembro_dezembro.groupby('nome_transacao')['valor_transacao'].sum()
res_mov_maio_agosto = df_mov_maio_agosto.groupby('nome_transacao')['valor_transacao'].sum()


print('Movimentações janeiro a abril')
print(res_mov_janeiro_abril)

print('Movimentações maio a agosto')
print(res_mov_maio_agosto)

print('Movimentações setembro a dezembro')
print(res_mov_setembro_dezembro)


Movimentações janeiro a abril
nome_transacao
Compra Crédito                       280618.95
Compra Débito                        210258.11
DOC - Realizado                      474420.00
DOC - Recebido                      3336730.00
Depósito em espécie                  122634.34
Estorno de Debito                      6320.99
Pagamento de boleto                  118293.80
Pix - Realizado                      824442.60
Pix - Recebido                       412889.30
Pix Saque                              2790.00
Saque                                154225.00
TED - Realizado                      983220.00
TED - Recebido                      5836662.00
Transferência entre CC - Crédito      48507.10
Transferência entre CC - Débito       37555.10
Name: valor_transacao, dtype: float64
Movimentações maio a agosto
nome_transacao
Compra Crédito                       312351.94
Compra Débito                        223835.05
DOC - Realizado                      528170.00
DOC - Recebido              